server/src/main/java/org/elasticsearch/indices/recovery/plan/SnapshotsRecoveryPlannerService.java

```diff
@@ -70,17 +75,38 @@ private ShardRecoveryPlan computeRecoveryPlanWithSnapshots(Store.MetadataSnapsho
 
         if (latestSnapshotOpt.isEmpty()) {
             // If we couldn't find any valid  snapshots, fallback to the source
-            return new ShardRecoveryPlan(ShardRecoveryPlan.SnapshotFilesToRecover.EMPTY,
-                filesMissingInTarget,
-                sourceTargetDiff.identical,
-                startingSeqNo,
-                translogOps,
-                sourceMetadata
-            );
+            return getRecoveryPlanUsingSourceNode(sourceMetadata, sourceTargetDiff, filesMissingInTarget, startingSeqNo, translogOps);
         }
 
         ShardSnapshot latestSnapshot = latestSnapshotOpt.get();
 
+        if (latestSnapshot.isLogicallyEquivalent(shardStateIdentifier)) {
+            Store.RecoveryDiff recoveryDiff = latestSnapshot.getMetadataSnapshot().recoveryDiff(sourceMetadata);
+
+            // Primary failed over after the snapshot was taken
+            if (recoveryDiff.different.isEmpty() == false || recoveryDiff.missing.isEmpty() == false) {
```

Comment:
I wonder if we should do some version checks here, since we're not relying on file contents being equal as we do in the regular path.